corehq/apps/geospatial/es.py

```diff
@@ -1,7 +1,115 @@
+from corehq.apps.es import filters
+from corehq.apps.es.aggregations import (
+    GeohashGridAggregation,
+    NestedAggregation,
+    TermsAggregation, FilterAggregation,
+)
+from corehq.apps.es.case_search import PROPERTY_GEOPOINT_VALUE, PROPERTY_KEY
+from corehq.apps.geospatial.const import MAX_GEOHASH_DOC_COUNT
 
-def get_geohashes(domain, *, case_property, coords=None, precision=None):
+AGG_NAME = 'geohashes'
+
+
+def find_precision(query, case_property):
+    """
+    Uses a binary search to find the value for ``precision`` that
+    maximizes geohash doc count but does not exceed
+    ``MAX_GEOHASH_DOC_COUNT``.
+
+    (A lower value of ``precision`` will increase the doc count, and a
+    higher value will decrease it.)
+    """
+    lower = 1
+    upper = 12
+    while True:
+        precision = mid(lower, upper)
+        max_doc_count = get_max_doc_count(query, case_property, precision)
+        if max_doc_count > MAX_GEOHASH_DOC_COUNT and precision > lower:
+            # Increase next value of precision to decrease doc count
+            lower = precision
+        elif max_doc_count < MAX_GEOHASH_DOC_COUNT and precision < upper:
+            # Decrease next value of precision to increase doc count
+            upper = precision
+        else:
+            # max_doc_count == MAX_GEOHASH_DOC_COUNT
+            # or upper and lower are equal or adjacent
+            return precision
+
+
+def get_max_doc_count(query, case_property, precision):
+    query = query.clone()
+    nested_agg = NestedAggregation('case_properties', CASE_PROPERTIES_PATH)
+    filter_ = filters.term(PROPERTY_KEY, case_property)
+    filter_agg = FilterAggregation('case_property', filter_)
+    query = query.aggregation(
+        nested_agg.aggregation(
+            filter_agg.aggregation(
+                GeohashGridAggregation(AGG_NAME, PROPERTY_GEOPOINT_VALUE, precision)
+            )
+        )
+    )
+    # Results in the following aggregations:
+    #     "aggs": {
+    #         "case_properties": {
+    #             "nested": {
+    #                 "path": "case_properties"
+    #             },
+    #             "aggs": {
+    #                 "case_property": {
+    #                     "filter": {
+    #                         "term": {
+    #                             "case_properties.key.exact": "gps_point"
+    #                         }
+    #                     },
+    #                     "aggs": {
+    #                         "geohashes": {
+    #                             "geohash_grid": {
+    #                                 "field": "case_properties.geopoint_value",
+    #                                 "precision": 4
+    #                             }
+    #                         }
+    #                     }
+    #                 }
+    #             }
+    #         }
+    #     }
+
+    queryset = query.run()  # ESError: TransportError(404, 'index_not_found_exception', 'no such index')
```

Comment:
@esoergel can you tell, by looking at the JSON above, why this results in `index_not_found_exception`? I have tried ...
* dropping the "nested" aggregation
* flipping the "nested" and "filter" aggregations around
* setting the "geohash_grind" field to "geopoint_value" without the "case_properties" path

... but I can't figure out why it can't find the (indexed) "geopoint_value" field.

Reply:
That looks like an issue with the index, not the mapping or filtering.  Like, where does `query` come from?  Does `query.run()` work with the rest of this function commented out?

Reply:
Oh, sorry! [`query` is defined here](https://github.com/dimagi/commcare-hq/pull/33540/files#diff-e55674d49765fb7a9c1304099b36b61b35fa6d1a93039d3e05c2bca49f9aeae4R113) in a report named "CaseGroupingReport". `query` is a `CaseSearchES` instance and at this point in the code it will have report filters applied. The report's `_build_query()` method calls the `find_precision()` function above, and `find_precision()` calls `get_max_doc_count()`. They do this to ensure that the aggregation returns buckets of no more that 10,000 cases. Happy to jump on a call for more context.

Reply:
looks like I missed you.  The error you're reporting looks like the index (the case search index) isn't present.  That is, it doesn't sound like it has anything to do with the JSON produced here.  Have you tried commenting out everything except `query.run()`?  I'd bet it's something like your local environment being broken (most plausible), or perhaps something weird in here messes with the query adapter.  Another thing to try is putting in some print statements to see what `query.adapter` is going in to this function and at the end.

Reply:
Looping back: After conversations with @esoergel , some additional eyeballs from the team, and some excellent debugging by @mkangia , unit tests are passing. I will close this draft PR, and open a normal PR.